File: baselib-java/src/main/java/com/kuky/baselib_java/generalWidget/ListHandlerView.java
package com.kuky.baselib_java.generalWidget;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.graphics.Color;
import android.graphics.drawable.AnimationDrawable;
import android.support.annotation.DrawableRes;
import android.support.annotation.NonNull;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.RelativeLayout;

import com.kuky.baselib_java.BR;
import com.kuky.baselib_java.R;
import com.kuky.baselib_java.baseAdapter.BaseRvHeaderFooterAdapter;
import com.kuky.baselib_java.databinding.ListHandlerBinding;
import com.scwang.smartrefresh.layout.SmartRefreshLayout;
import com.scwang.smartrefresh.layout.listener.OnLoadmoreListener;
import com.scwang.smartrefresh.layout.listener.OnRefreshListener;

import java.util.List;

/**
 * @author Kuky
 */
public class ListHandlerView<T> extends RelativeLayout {
    public static final int LOAD_LOADING_STATE = 0x00010000;
    public static final int LOAD_EMPTY_STATE = 0x00010001;
    public static final int LOAD_FAILED_STATE = 0x00010002;
    public static final int LOAD_SUCCEED_STATE = 0x00010003;
    public static final int HORIZONTAL_PROGRESS = 0x00020000;
    public static final int CIRCLE_PROGRESS = 0x00020001;

    private ListHandlerBinding mHandlerBinding;
    private int mListState = LOAD_LOADING_STATE;
    private int mProgressStyle = HORIZONTAL_PROGRESS;
    private OnListReloadListener mOnListReloadListener;
    private BaseRvHeaderFooterAdapter mListAdapter;
    private List<T> mData;

    /**
     * 重新加载监听
     */
    public interface OnListReloadListener {
        void reLoad();
    }

    public void setOnListReloadListener(OnListReloadListener loadListener) {
        this.mOnListReloadListener = loadListener;
    }

    /**
     * 刷新监听
     *
     * @param refreshListener
     */
    public void setRefreshListener(OnRefreshListener refreshListener) {
        if (refreshListener != null) {
            mHandlerBinding.listPage.setOnRefreshListener(refreshListener);
        }
    }

    /**
     * 加载更多监听
     *
     * @param loadMoreListener
     */
    public void setLoadMoreListener(OnLoadmoreListener loadMoreListener) {
        if (loadMoreListener != null) {
            mHandlerBinding.listPage.setOnLoadmoreListener(loadMoreListener);
        }
    }

    public ListHandlerView(Context context) {
        this(context, null);
    }

    public ListHandlerView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public ListHandlerView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initHandlerView(context);
    }

    private void initHandlerView(Context context) {
        mHandlerBinding = DataBindingUtil.inflate(LayoutInflater.from(context), R.layout.list_handler, this, true);
        mHandlerBinding.setVariable(BR.listHandle, this);
        hideAllPage();
        switch (mProgressStyle) {
            case HORIZONTAL_PROGRESS:
                mHandlerBinding.loadingPage.setImageResource(R.drawable.ic_horizontal_loading_progress);
                mHandlerBinding.loadingPage.setBackgroundColor(Color.TRANSPARENT);
                break;
            case CIRCLE_PROGRESS:
                mHandlerBinding.loadingPage.setImageResource(R.drawable.ic_circle_loading_progress);
                mHandlerBinding.loadingPage.setBackgroundColor(Color.parseColor("#75000000"));
                break;
        }

        switch (mListState) {
            case LOAD_LOADING_STATE:
                mHandlerBinding.loadingPage.setVisibility(VISIBLE);
                ((AnimationDrawable) mHandlerBinding.loadingPage.getDrawable()).start();
                break;
            case LOAD_EMPTY_STATE:
                mHandlerBinding.nothingPage.setVisibility(VISIBLE);
                break;
            case LOAD_FAILED_STATE:
                mHandlerBinding.reloadPage.setVisibility(VISIBLE);
                break;
            case LOAD_SUCCEED_STATE:
                mHandlerBinding.listPage.setVisibility(VISIBLE);
                break;
        }
    }

    private void hideAllPage() {
        mHandlerBinding.loadingPage.setVisibility(GONE);
        mHandlerBinding.nothingPage.setVisibility(GONE);
        mHandlerBinding.reloadPage.setVisibility(GONE);
        mHandlerBinding.listPage.setVisibility(GONE);
    }

    /**
     * 设置加载界面
     *
     * @param style
     * @return
     */
    public ListHandlerView setLoadingPage(int style) {
        this.mProgressStyle = style;
        switch (mProgressStyle) {
            case HORIZONTAL_PROGRESS:
                mHandlerBinding.loadingPage.setImageResource(R.drawable.ic_horizontal_loading_progress);
                mHandlerBinding.loadingPage.setBackgroundColor(Color.TRANSPARENT);
                break;
            case CIRCLE_PROGRESS:
                mHandlerBinding.loadingPage.setImageResource(R.drawable.ic_circle_loading_progress);
                mHandlerBinding.loadingPage.setBackgroundColor(Color.parseColor("#75000000"));
                break;
        }
        ((AnimationDrawable) mHandlerBinding.loadingPage.getDrawable()).start();
        return this;
    }

    public void replaceLoadingProgress(@DrawableRes int drawableRes) {
        mHandlerBinding.loadingPage.setImageResource(drawableRes);
        ((AnimationDrawable) mHandlerBinding.loadingPage.getDrawable()).start();
    }

    public void replaceLoadingProgress(AnimationDrawable drawable) {
        mHandlerBinding.loadingPage.setImageDrawable(drawable);
        drawable.start();
    }

    /**
     * 设置加载失败界面图片
     *
     * @param drawableRes
     * @return
     */
    public void setEmptyDrawable(@DrawableRes int drawableRes) {
        mHandlerBinding.nothingPage.setImageResource(drawableRes);
    }

    /**
     * 设置无数据界面图片
     *
     * @param drawableRes
     * @return
     */
    public void setFailedDrawable(@DrawableRes int drawableRes) {
        mHandlerBinding.reloadPage.setImageResource(drawableRes);
    }

    /**
     * 列表界面
     *
     * @param adapter
     * @param listener
     * @param rvManager
     * @param anim
     * @param decoration
     * @return
     */
    public void setLoadSuccessPage(@NonNull BaseRvHeaderFooterAdapter adapter, boolean showBackTop,
                                   BaseRvHeaderFooterAdapter.OnItemClickListener listener,
                                   @NonNull final RecyclerView.LayoutManager rvManager,
                                   RecyclerView.ItemAnimator anim,
                                   RecyclerView.ItemDecoration decoration) {
        this.mListAdapter = adapter;
        this.mData = adapter.getAdapterData();
        mHandlerBinding.listContainer.setAdapter(mListAdapter);
        mHandlerBinding.listContainer.setLayoutManager(rvManager);

        if (showBackTop) {
            getListContainer().addOnScrollListener(new RecyclerView.OnScrollListener() {
                @Override
                public void onScrollStateChanged(RecyclerView recyclerView, int newState) {
                    super.onScrollStateChanged(recyclerView, newState);
                    if (rvManager instanceof LinearLayoutManager) {
                        int firstVisibleItemPosition = ((LinearLayoutManager) rvManager).findFirstVisibleItemPosition();
                        switch (newState) {
                            case RecyclerView.SCROLL_STATE_IDLE:
                                mHandlerBinding.backTop.setVisibility(firstVisibleItemPosition <= 2 ? View.GONE : View.VISIBLE);
                                break;
                            case RecyclerView.SCROLL_STATE_DRAGGING:
                                mHandlerBinding.backTop.setVisibility(View.GONE);
                                break;
                        }
                    } else if (rvManager instanceof GridLayoutManager) {
                        int firstVisibleItemPosition = ((GridLayoutManager) rvManager).findFirstVisibleItemPosition();
                        switch (newState) {
                            case RecyclerView.SCROLL_STATE_IDLE:
                                mHandlerBinding.backTop.setVisibility(firstVisibleItemPosition <= 6 ? View.GONE : View.VISIBLE);
                                break;
                            case RecyclerView.SCROLL_STATE_DRAGGING:
                                mHandlerBinding.backTop.setVisibility(GONE);
                                break;
                        }
                    } else if (rvManager instanceof StaggeredGridLayoutManager) {
                        int[] firstVisibleItemPositions = new int[1];
                        ((StaggeredGridLayoutManager) rvManager).findFirstVisibleItemPositions(firstVisibleItemPositions);
                        switch (newState) {
                            case RecyclerView.SCROLL_STATE_IDLE:
                                mHandlerBinding.backTop.setVisibility(firstVisibleItemPositions[0] <= 6 ? View.GONE : View.VISIBLE);
                                break;
                            case RecyclerView.SCROLL_STATE_DRAGGING:
                                mHandlerBinding.backTop.setVisibility(GONE);
                                break;
                        }
                    }
                }
            });
        } else {
            mHandlerBinding.backTop.setVisibility(GONE);
        }

        if (listener != null) {
            mListAdapter.setOnItemClickListener(listener);
        }

        mHandlerBinding.listContainer.setItemAnimator(anim == null ? new DefaultItemAnimator() : anim);

        if (decoration != null) {
            mHandlerBinding.listContainer.addItemDecoration(decoration);
        }
    }

    public void setListAnim(@NonNull RecyclerView.ItemAnimator anim) {
        mHandlerBinding.listContainer.setItemAnimator(anim);
    }

    public void setListDecoration(@NonNull RecyclerView.ItemDecoration decoration) {
        mHandlerBinding.listContainer.addItemDecoration(decoration);
    }

    public void setBackTopButtonResource(@DrawableRes int drawableRes) {
        mHandlerBinding.backTop.setImageResource(drawableRes);
    }

    public void addHeader(ViewDataBinding headerBinding) {
        mListAdapter.addHeaderBinding(headerBinding);
    }

    public void addFooter(ViewDataBinding footerBinding) {
        mListAdapter.addFooterBinding(footerBinding);
    }

    public List<ViewDataBinding> getHeaders() {
        return mListAdapter.getHeaderBindings();
    }

    public List<ViewDataBinding> getFooters() {
        return mListAdapter.getFooterBindings();
    }

    public void reloadClick(View view) {
        if (mOnListReloadListener != null && mListState == LOAD_FAILED_STATE) {
            mOnListReloadListener.reLoad();
        }
    }

    public void backTop(View view) {
        getListContainer().smoothScrollToPosition(0);
    }

    /**
     * 更新界面数据
     *
     * @param data
     */
    public void updateData(List<T> data) {
        this.mData = data;
        this.mListAdapter.updateAdapterData(mData);
    }

    public List<T> getListData() {
        return mListAdapter.getAdapterData();
    }

    /**
     * 添加数据
     *
     * @param t
     */
    public void addData(T t) {
        this.mListAdapter.addData(mData);
    }

    /**
     * 添加数据
     *
     * @param data
     */
    public void addData(List<T> data) {
        this.mListAdapter.addAllData(mData);
    }

    /**
     * 去除特定数据
     *
     * @param t
     */
    public void removeData(T t) {
        mListAdapter.removeData(mData);
    }

    /**
     * 去除特定数据
     *
     * @param position
     */
    public void removeData(int position) {
        mListAdapter.removeData(mData);
    }

    /**
     * 清除所有数据
     */
    public void cleanData() {
        mListAdapter.cleanData();
    }

    /**
     * 重置状态
     *
     * @param state
     */
    public void setViewState(int state) {
        this.mListState = state;
        hideAllPage();
        switch (mListState) {
            case LOAD_LOADING_STATE:
                mHandlerBinding.loadingPage.setVisibility(VISIBLE);
                ((AnimationDrawable) mHandlerBinding.loadingPage.getDrawable()).start();
                break;
            case LOAD_EMPTY_STATE:
                mHandlerBinding.nothingPage.setVisibility(VISIBLE);
                break;
            case LOAD_FAILED_STATE:
                mHandlerBinding.reloadPage.setVisibility(VISIBLE);
                break;
            case LOAD_SUCCEED_STATE:
                mHandlerBinding.listPage.setVisibility(VISIBLE);
                break;
        }
    }

    public SmartRefreshLayout getRefreshLoadContainer() {
        return mHandlerBinding.listPage;
    }

    public RecyclerView getListContainer() {
        return mHandlerBinding.listContainer;
    }

    public void enabledRefresh(boolean enabledRefresh) {
        getRefreshLoadContainer().setEnableRefresh(enabledRefresh);
    }

    public void enabledLoadMore(boolean enabledLoadMore) {
        getRefreshLoadContainer().setEnableLoadmore(enabledLoadMore);
    }

    public void enabledAutoLoadMore(boolean enabledAutoLoadMore) {
        getRefreshLoadContainer().setEnableLoadmore(enabledAutoLoadMore);
    }

    public void enabledLoadMoreWhenNotFull(boolean enabledLoadMoreWhenNotFull) {
        getRefreshLoadContainer().setEnableLoadmoreWhenContentNotFull(enabledLoadMoreWhenNotFull);
    }

    public void stopLoadMore() {
        getRefreshLoadContainer().setLoadmoreFinished(true);
    }

    public void restartLoadMore() {
        getRefreshLoadContainer().setLoadmoreFinished(false);
    }

    public void finishRefresh() {
        getRefreshLoadContainer().finishRefresh();
    }

    public void finishLoadMore() {
        getRefreshLoadContainer().finishLoadmore();
    }
}

File: baselib/src/main/java/com/kuky/baselib/generalWidget/ListHandlerView.kt
package com.kuky.baselib.generalWidget

import android.content.Context
import android.databinding.DataBindingUtil
import android.databinding.ViewDataBinding
import android.graphics.Color
import android.graphics.drawable.AnimationDrawable
import android.support.annotation.DrawableRes
import android.support.v7.widget.*
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.RelativeLayout
import com.kuky.baselib.R
import com.kuky.baselib.baseAdapter.BaseRvHeaderFooterAdapter
import com.kuky.baselib.databinding.ListHandlerBinding
import com.scwang.smartrefresh.layout.SmartRefreshLayout
import com.scwang.smartrefresh.layout.listener.OnLoadmoreListener
import com.scwang.smartrefresh.layout.listener.OnRefreshListener

/**
 * @author Kuky
 */
class ListHandlerView<T : Any> : RelativeLayout {
    private lateinit var mHandlerBinding: ListHandlerBinding
    private var mState = LOAD_LOADING_STATE
    private var mProgressStyle = HORIZONTAL_PROGRESS
    private var mListAdapter: BaseRvHeaderFooterAdapter<T, ViewDataBinding>? = null
    private var mData: MutableList<T>? = null
    private var mOnListReloadListener: OnListReloadListener? = null

    interface OnListReloadListener {
        fun onListReload()
    }

    fun setOnListReloadListener(listener: OnListReloadListener) {
        this.mOnListReloadListener = listener
    }

    fun setOnRefreshListener(listener: OnRefreshListener) {
        mHandlerBinding.listPage.setOnRefreshListener(listener)
    }

    fun setOnLoadMoreListener(listener: OnLoadmoreListener) {
        mHandlerBinding.listPage.setOnLoadmoreListener(listener)
    }

    constructor(context: Context) : this(context, null)

    constructor(context: Context, attr: AttributeSet?) : this(context, attr, 0)

    constructor(context: Context, attr: AttributeSet?, defStyleAttr: Int) : super(context, attr, defStyleAttr) {
        initHandlerView(context)
    }

    private fun initHandlerView(context: Context) {
        mHandlerBinding = DataBindingUtil.inflate(LayoutInflater.from(context), R.layout.list_handler, this, true)
        mHandlerBinding.listHandle = this@ListHandlerView
        hideAllPage()

        when (mProgressStyle) {
            HORIZONTAL_PROGRESS -> mHandlerBinding.loadingPage.setImageResource(R.drawable.ic_horizontal_loading_progress)
            CIRCLE_PROGRESS -> {
                mHandlerBinding.loadingPage.setImageResource(R.drawable.ic_circle_loading_progress)
                mHandlerBinding.loadingPage.setBackgroundColor(Color.BLACK)
            }
        }

        when (mState) {
            LOAD_LOADING_STATE -> {
                mHandlerBinding.loadingPage.visibility = View.VISIBLE
                (mHandlerBinding.loadingPage.drawable as AnimationDrawable).start()
            }
            LOAD_EMPTY_STATE -> mHandlerBinding.nothingPage.visibility = View.VISIBLE
            LOAD_FAILED_STATE -> mHandlerBinding.reloadPage.visibility = View.VISIBLE
            LOAD_SUCCEED_STATE -> mHandlerBinding.listPage.visibility = View.VISIBLE
        }
    }

    private fun hideAllPage() {
        mHandlerBinding.loadingPage.visibility = GONE
        mHandlerBinding.reloadPage.visibility = GONE
        mHandlerBinding.nothingPage.visibility = GONE
        mHandlerBinding.listPage.visibility = GONE
    }

    fun setLoadingStyle(style: Int) {
        this.mProgressStyle = style
        when (mProgressStyle) {
            HORIZONTAL_PROGRESS -> mHandlerBinding.loadingPage.setImageResource(R.drawable.ic_horizontal_loading_progress)
            CIRCLE_PROGRESS -> mHandlerBinding.loadingPage.setImageResource(R.drawable.ic_circle_loading_progress)
        }
        (mHandlerBinding.loadingPage.drawable as AnimationDrawable).start()
    }

    fun replaceLoadingProgress(drawableRes: Int) {
        mHandlerBinding.loadingPage.setImageResource(drawableRes)
        (mHandlerBinding.loadingPage.drawable as AnimationDrawable).start()
    }

    fun repalceLoadingProgress(drawable: AnimationDrawable) {
        mHandlerBinding.loadingPage.setImageDrawable(drawable)
        drawable.start()
    }

    fun setEmptyDrawable(@DrawableRes drawableRes: Int) {
        mHandlerBinding.nothingPage.setImageResource(drawableRes)
    }

    fun setFailedDrawable(@DrawableRes drawableRes: Int) {
        mHandlerBinding.reloadPage.setImageResource(drawableRes)
    }

    fun setListPages(adapter: BaseRvHeaderFooterAdapter<T, ViewDataBinding>, showBackTop: Boolean,
                     listener: BaseRvHeaderFooterAdapter.OnItemClickListener?,
                     rvManager: RecyclerView.LayoutManager,
                     anim: RecyclerView.ItemAnimator?,
                     decoration: RecyclerView.ItemDecoration?) {

        this.mListAdapter = adapter
        this.mData = adapter.getAdapterData()

        mHandlerBinding.listContainer.adapter = mListAdapter

        mHandlerBinding.listContainer.layoutManager = rvManager

        /**
         * RecyclerView 返回顶部按钮
         */
        if (showBackTop) {
            getListContainer().addOnScrollListener(object : RecyclerView.OnScrollListener() {
                override fun onScrollStateChanged(recyclerView: RecyclerView?, newState: Int) {
                    super.onScrollStateChanged(recyclerView, newState)
                    when (rvManager) {
                        is LinearLayoutManager -> {
                            val firstVisibleItemPosition = rvManager.findFirstVisibleItemPosition()
                            when (newState) {
                                RecyclerView.SCROLL_STATE_IDLE -> {
                                    mHandlerBinding.backTop.visibility = if (firstVisibleItemPosition <= 3) View.GONE else View.VISIBLE
                                }

                                RecyclerView.SCROLL_STATE_DRAGGING -> mHandlerBinding.backTop.visibility = View.GONE
                            }
                        }

                        is GridLayoutManager -> {
                            val firstVisibleItemPosition = rvManager.findFirstVisibleItemPosition()
                            when (newState) {
                                RecyclerView.SCROLL_STATE_IDLE -> {
                                    mHandlerBinding.backTop.visibility = if (firstVisibleItemPosition <= 6) View.GONE else View.VISIBLE
                                }

                                RecyclerView.SCROLL_STATE_DRAGGING -> mHandlerBinding.backTop.visibility = View.GONE
                            }
                        }

                        is StaggeredGridLayoutManager -> {
                            val firstVisibleItemPositions = IntArray(0)
                            rvManager.findFirstVisibleItemPositions(firstVisibleItemPositions)
                            when (newState) {
                                RecyclerView.SCROLL_STATE_IDLE -> {
                                    mHandlerBinding.backTop.visibility = if (firstVisibleItemPositions[0] <= 6) View.GONE else View.VISIBLE
                                }

                                RecyclerView.SCROLL_STATE_DRAGGING -> mHandlerBinding.backTop.visibility = View.GONE
                            }
                        }
                    }
                }
            })
        } else
            mHandlerBinding.backTop.visibility = View.GONE

        if (listener != null) {
            mListAdapter!!.setOnItemClickListener(listener)
        }

        mHandlerBinding.listContainer.itemAnimator = anim ?: DefaultItemAnimator()

        if (decoration != null) {
            mHandlerBinding.listContainer.addItemDecoration(decoration)
        }
    }

    fun setListAnim(anim: RecyclerView.ItemAnimator) {
        mHandlerBinding.listContainer.itemAnimator = anim
    }

    fun setListDecoration(decoration: RecyclerView.ItemDecoration) {
        mHandlerBinding.listContainer.addItemDecoration(decoration)
    }

    fun setBackTopButtonResource(resId: Int) {
        mHandlerBinding.backTop.setBackgroundResource(resId)
    }

    fun addHeader(headerBinding: ViewDataBinding) {
        mListAdapter!!.addHeaderBinding(headerBinding)
    }

    fun addFooter(footerBinding: ViewDataBinding) {
        mListAdapter!!.addFooterBinding(footerBinding)
    }

    fun getHeaders(): MutableList<ViewDataBinding> {
        return mListAdapter!!.getHeaderBindings()
    }

    fun getFooters(): MutableList<ViewDataBinding> {
        return mListAdapter!!.getFooterBindings()
    }

    fun reloadClick(view: View) {
        if (mOnListReloadListener != null && mState == LOAD_FAILED_STATE) {
            mOnListReloadListener!!.onListReload()
        }
    }

    fun backTop(view: View) {
        getListContainer().smoothScrollToPosition(0)
    }

    fun updateData(data: MutableList<T>?) {
        this.mListAdapter!!.updateAdapterData(data)
    }

    fun getListData(): MutableList<T>? {
        return mListAdapter!!.getAdapterData()
    }

    fun addData(data: T) {
        this.mListAdapter!!.addData(data)
    }

    fun addAllData(data: MutableList<T>) {
        this.mListAdapter!!.addAllData(data)
    }

    fun removeData(data: T) {
        this.mListAdapter!!.removeData(data)
    }

    fun removeData(position: Int) {
        this.mListAdapter!!.removeData(position)
    }

    fun cleanData() {
        this.mListAdapter!!.clearData()
    }

    fun setHandlerState(state: Int) {
        this.mState = state
        hideAllPage()
        when (mState) {
            LOAD_LOADING_STATE -> {
                mHandlerBinding.loadingPage.visibility = View.VISIBLE
                (mHandlerBinding.loadingPage.drawable as AnimationDrawable).start()
            }
            LOAD_EMPTY_STATE -> mHandlerBinding.nothingPage.visibility = View.VISIBLE
            LOAD_FAILED_STATE -> mHandlerBinding.reloadPage.visibility = View.VISIBLE
            LOAD_SUCCEED_STATE -> mHandlerBinding.listPage.visibility = View.VISIBLE
        }
    }

    fun getRefreshLoadContainer(): SmartRefreshLayout {
        return mHandlerBinding.listPage
    }

    fun getListContainer(): RecyclerView {
        return mHandlerBinding.listContainer
    }

    fun enabledRefresh(enabledRefresh: Boolean = true) {
        getRefreshLoadContainer().isEnableRefresh = enabledRefresh
    }

    fun enabledLoadMore(enabledLoadMore: Boolean = true) {
        getRefreshLoadContainer().isEnableLoadmore = enabledLoadMore
    }

    fun enabledAutoRefresh(enabledAutoRefresh: Boolean = false) {
        getRefreshLoadContainer().isEnableAutoLoadmore = enabledAutoRefresh
    }

    fun enabledAutoLoadMore(enabledAutoLoadMore: Boolean = false) {
        getRefreshLoadContainer().isEnableAutoLoadmore = enabledAutoLoadMore
    }

    fun enabledLoadMoreWhenNotFull(enabledLoadMoreWhenNotFull: Boolean = false) {
        getRefreshLoadContainer().setEnableLoadmoreWhenContentNotFull(enabledLoadMoreWhenNotFull)
    }

    fun stopLoadMore() {
        getRefreshLoadContainer().isLoadmoreFinished = true
    }

    fun restartLoadMore() {
        getRefreshLoadContainer().isLoadmoreFinished = false
    }

    fun finishRefresh() {
        getRefreshLoadContainer().finishRefresh()
    }

    fun finishLoadMore() {
        getRefreshLoadContainer().finishLoadmore()
    }

    companion object {
        val LOAD_LOADING_STATE = 0x00010000
        val LOAD_EMPTY_STATE = 0x00010001
        val LOAD_FAILED_STATE = 0x00010002
        val LOAD_SUCCEED_STATE = 0x00010003
        val HORIZONTAL_PROGRESS = 0x00020000
        val CIRCLE_PROGRESS = 0x00020001
    }
}
